File: skills/optimization/reference/src/gradient-descent.ts
/**
 * Gradient descent (steepest descent) optimization.
 *
 * The simplest gradient-based method: d = -gradient, with backtracking line search.
 * Serves as a baseline for comparison with BFGS and L-BFGS.
 *
 * @node gradient-descent
 * @depends-on vec-ops, result-types, line-search, finite-diff
 * @contract gradient-descent.test.ts
 * @hint performance: Gradient descent is slow on ill-conditioned problems (like
 *       Rosenbrock). BFGS is almost always better. Include this node for
 *       completeness and as a teaching example.
 * @hint line-search: Uses backtracking (Armijo) by default, not Wolfe.
 *       GD doesn't maintain a Hessian approximation, so Wolfe isn't needed.
 */

import { negate, normInf, sub, addScaled } from "./vec-ops";
import {
  type OptimizeResult,
  type OptimizeOptions,
  defaultOptions,
  checkConvergence,
  isConverged,
  convergenceMessage,
} from "./result-types";
import { backtrackingLineSearch } from "./line-search";
import { forwardDiffGradient } from "./finite-diff";

/**
 * Minimize a function using gradient descent with backtracking line search.
 *
 * @provenance Standard steepest descent algorithm (Cauchy 1847)
 * @provenance Backtracking line search from Nocedal & Wright, Algorithm 3.1
 */
export function gradientDescent(
  f: (x: number[]) => number,
  x0: number[],
  grad?: (x: number[]) => number[],
  options?: Partial<OptimizeOptions>,
): OptimizeResult {
  const opts = defaultOptions(options);
  const gradFn = grad ?? ((x: number[]) => forwardDiffGradient(f, x));

  let x = x0.slice();
  let fx = f(x);
  let gx = gradFn(x);
  let functionCalls = 1;
  let gradientCalls = 1;

  // Check if already at a minimum
  let gradNorm = normInf(gx);
  const initialCheck = checkConvergence(gradNorm, Infinity, Infinity, 0, opts);
  if (initialCheck && isConverged(initialCheck)) {
    return {
      x, fun: fx, gradient: gx.slice(),
      iterations: 0, functionCalls, gradientCalls,
      converged: true, message: convergenceMessage(initialCheck),
    };
  }

  for (let iteration = 1; iteration <= opts.maxIterations; iteration++) {
    // Search direction: steepest descent
    const d = negate(gx);

    // Backtracking line search
    const ls = backtrackingLineSearch(f, x, d, fx, gx);
    functionCalls += ls.functionCalls;

    if (!ls.success) {
      return {
        x, fun: fx, gradient: gx.slice(),
        iterations: iteration, functionCalls, gradientCalls,
        converged: false,
        message: "Stopped: line search failed to find acceptable step",
      };
    }

    const xNew = addScaled(x, d, ls.alpha);
    const fNew = ls.fNew;
    const gNew = gradFn(xNew);
    gradientCalls++;

    const stepNorm = normInf(sub(xNew, x));
    const funcChange = Math.abs(fNew - fx);
    gradNorm = normInf(gNew);

    x = xNew;
    fx = fNew;
    gx = gNew;

    // Check convergence
    const reason = checkConvergence(gradNorm, stepNorm, funcChange, iteration, opts);
    if (reason) {
      return {
        x: x.slice(), fun: fx, gradient: gx.slice(),
        iterations: iteration, functionCalls, gradientCalls,
        converged: isConverged(reason), message: convergenceMessage(reason),
      };
    }
  }

  return {
    x: x.slice(), fun: fx, gradient: gx.slice(),
    iterations: opts.maxIterations, functionCalls, gradientCalls,
    converged: false,
    message: `Stopped: reached maximum iterations (${opts.maxIterations})`,
  };
}

File: research/experiments/sweep/runs/opt-full-csharp-sonnet45/src/GradientDescent.cs
namespace Optimization;

public static class GradientDescent
{
    public static OptimizeResult Minimize(
        Func<double[], double> f, double[] x0,
        Func<double[], double[]>? grad = null, OptimizeOptions? options = null)
    {
        var opts = ResultTypes.MergeOptions(options);
        var gradFn = grad ?? (x => FiniteDiff.ForwardDiffGradient(f, x));

        double[] x = (double[])x0.Clone();
        double fx = f(x);
        double[] gx = gradFn(x);
        int functionCalls = 1;
        int gradientCalls = 1;

        double gradNorm = VecOps.NormInf(gx);
        if (gradNorm <= opts.GradTol)
        {
            return new OptimizeResult
            {
                X = x, Fun = fx, Gradient = (double[])gx.Clone(),
                Iterations = 0, FunctionCalls = functionCalls, GradientCalls = gradientCalls,
                Converged = true, Message = "Convergence: gradient norm below tolerance"
            };
        }

        for (int iter = 1; iter <= opts.MaxIterations; iter++)
        {
            double[] d = VecOps.Negate(gx);

            var ls = LineSearch.BacktrackingLineSearch(f, x, d, fx, gx);
            functionCalls += ls.FunctionCalls;

            if (!ls.Success)
            {
                return new OptimizeResult
                {
                    X = x, Fun = fx, Gradient = (double[])gx.Clone(),
                    Iterations = iter, FunctionCalls = functionCalls, GradientCalls = gradientCalls,
                    Converged = false, Message = "Stopped: line search failed"
                };
            }

            double[] xNew = VecOps.AddScaled(x, d, ls.Alpha);
            double fNew = ls.FNew;
            double[] gNew = gradFn(xNew);
            gradientCalls++;

            double stepNorm = VecOps.NormInf(VecOps.Sub(xNew, x));
            double funcChange = Math.Abs(fNew - fx);

            x = xNew; fx = fNew; gx = gNew;
            gradNorm = VecOps.NormInf(gx);

            var reason = ResultTypes.CheckConvergence(gradNorm, stepNorm, funcChange, iter, opts);
            if (reason != null)
            {
                return new OptimizeResult
                {
                    X = (double[])x.Clone(), Fun = fx, Gradient = (double[])gx.Clone(),
                    Iterations = iter, FunctionCalls = functionCalls, GradientCalls = gradientCalls,
                    Converged = ResultTypes.IsConverged(reason), Message = ResultTypes.ConvergenceMessage(reason)
                };
            }
        }

        return new OptimizeResult
        {
            X = (double[])x.Clone(), Fun = fx, Gradient = (double[])gx.Clone(),
            Iterations = opts.MaxIterations, FunctionCalls = functionCalls, GradientCalls = gradientCalls,
            Converged = false, Message = "Stopped: reached maximum iterations"
        };
    }
}
